meta/autoid/autoid.go

```diff
@@ -438,35 +492,81 @@ func NewAllocatorsFromTblInfo(store kv.Storage, schemaID int64, tblInfo *model.T
-// increment and offset at this time now. To simplify the rule is like (ID - offset) % increment = 0,
-// so the first autoID should be 9, then add increment to it to get 13.
-func (alloc *allocator) Alloc(tableID int64, n uint64, increment, offset int64) (int64, int64, error) {
+// The values(ID) returned by the IDIterator satisfy the equation `(ID - offset) % increment = 0`.
+func (alloc *allocator) Alloc(tableID int64, n uint64, increment, offset int64) (IDIterator, error) {
 	if tableID == 0 {
-		return 0, 0, errInvalidTableID.GenWithStackByArgs("Invalid tableID")
+		return IDIterator{}, errInvalidTableID.GenWithStackByArgs("Invalid tableID")
 	}
 	if n == 0 {
-		return 0, 0, nil
+		return IDIterator{}, nil
 	}
-	if alloc.allocType == AutoIncrementType || alloc.allocType == RowIDAllocType {
+	switch alloc.allocType {
+	case AutoIncrementType, RowIDAllocType, AutoRandomType:
 		if !validIncrementAndOffset(increment, offset) {
-			return 0, 0, errInvalidIncrementAndOffset.GenWithStackByArgs(increment, offset)
+			return IDIterator{}, errInvalidIncrementAndOffset.GenWithStackByArgs(increment, offset)
 		}
 	}
 	alloc.mu.Lock()
 	defer alloc.mu.Unlock()
-	if alloc.isUnsigned {
-		return alloc.alloc4Unsigned(tableID, n, increment, offset)
+	return alloc.alloc(tableID, n, increment, offset)
+}
+
+func (alloc *allocator) alloc(tableID int64, n uint64, increment, offset int64) (IDIterator, error) {
+	// Check offset rebase if necessary.
+	if alloc.cmp(offset-1, alloc.base).is(greater) {
+		if err := alloc.rebase(tableID, offset-1, true); err != nil {
+			return IDIterator{}, err
+		}
+	}
+	// calcNeededBatchSize calculates the total batch size needed.
+	batchSize, expectedEnd, isOverflow := alloc.calcNeededBatchSize(alloc.base, int64(n), increment, offset)
+	if isOverflow {
+		return IDIterator{}, ErrAutoincReadFailed
+	}
+	// The local rest is not enough for alloc, skip it.
+	if alloc.cmp(expectedEnd, alloc.end).is(greater) {
+		var newBase, newEnd int64
+		startTime := time.Now()
+		alloc.adjustStep(startTime)
+		err := kv.RunInNewTxn(alloc.store, true, func(txn kv.Transaction) error {
+			m := meta.NewMeta(txn)
+			var err1 error
+			newBase, err1 = getAutoIDByAllocType(m, alloc.dbID, tableID, alloc.allocType)
+			if err1 != nil {
+				return err1
+			}
+			batchSizeIsOutdated := newBase != alloc.base
+			if batchSizeIsOutdated {
+				batchSize, _, isOverflow = alloc.calcNeededBatchSize(newBase, int64(n), increment, offset)
+				if isOverflow {
+					return ErrAutoincReadFailed
+				}
+			}
+			actualStep := alloc.calcActualStep(newBase, batchSize)
+			newEnd, err1 = generateAutoIDByAllocType(m, alloc.dbID, tableID, actualStep, alloc.allocType)
+			return err1
+		})
+		metrics.AutoIDHistogram.WithLabelValues(metrics.TableAutoIDAlloc, metrics.RetLabel(err)).Observe(time.Since(startTime).Seconds())
+		if err != nil {
+			return IDIterator{}, err
+		}
+		alloc.lastAllocTime = time.Now()
```

Comment:
better update `alloc.step` here like `alloc.lastAllocTime` does if it's not customized. 
it's used to compute the next dynamic step.
```
if !alloc.customStep {
	alloc.step = nextStep
}
```

Reply:
`alloc.adjustStep()` updated the `alloc.step`:
```go
func (alloc *allocator) adjustStep(startTime time.Time) {
	if alloc.customStep {
		return
	}
	consumeDur := startTime.Sub(alloc.lastAllocTime)
	alloc.step = NextStep(alloc.step, consumeDur)
}
```

Reply:
I mean, the step will also be modified in L541`actualStep`
Maybe we should record the actualStep to `alloc.step` if it's not user-customized.

if !alloc.customStep {
	alloc.step = actualStep
}